### DISCUSSION TITLE
Wix 4 Harvesting Question

### DISCUSSION THREAD

**Discussion by MaroisConsulting** [wixtoolset/issues#7805] [Questions]
Here's my Product.wsx:
```
<Wix xmlns="http://wixtoolset.org/schemas/v4/wxs">
	
  <Package Name="ACME WIdgets" 
		   Manufacturer="Boyde, Dewey, Cheatum, and Howe" 
		   Version="1.0.0.0" 
		   UpgradeCode="86dca32d-c567-40d0-8bf3-8b56b94f3074">
	  
    <MajorUpgrade DowngradeErrorMessage="!(loc.DowngradeError)" />

	<MediaTemplate EmbedCab="yes" />
		  
    <Feature Id="Main">
      <ComponentGroupRef Id="AppComponents" />
    </Feature>
	  
  </Package>
	
</Wix>
```
I have this in my Wix project:
```
<Project Sdk="WixToolset.Sdk/4.0.2">
  <ItemGroup>
    <PackageReference Include="WixToolset.Heat" Version="4.0.2" />
    <PackageReference Include="WixToolset.UI.wixext" Version="4.0.2" />
    <PackageReference Include="WixToolset.Util.wixext" Version="4.0.2" />
    <PackageReference Include="WixToolset.VisualStudio.wixext" Version="4.0.2" />
  </ItemGroup>
  <ItemGroup>
    <ProjectReference Include="..\RDS.PaymentProcessing.UI.Desktop\RDS.PaymentProcessing.UI.Desktop.csproj" />
  </ItemGroup>
  
	<ItemGroup>
		
        <HarvestDirectory Include="..\RDS.PaymentProcessing.UI.Desktop\Bin\Release">
            <ComponentGroupName>AppComponents</ComponentGroupName>
            <DirectoryRefId>INSTALLFOLDER</DirectoryRefId>
            <SuppressRootDirectory>true</SuppressRootDirectory>
        </HarvestDirectory>
		
        <BindPath Include="..\RDS.PaymentProcessing.UI.Desktop\Bin\Release" />
		
  </ItemGroup>

</Project>

```
1. When I run, nothing is generated. Isn't there supposed to be a generated file with the harvest results in it? 

2. Also, there is no .MSI created anywhere that I can see.
3. Could someone point me to the documentation for this? I got this code snippet off of SO. I can't seem to find anything like it in these docs: https://www.firegiant.com/docs/wix/tutorial/

**Comment by robmen** [comment]
The problem is probably because you are using the wrong extension for your source code. It should be `.wxs`.

Just kidding. Your extension is wrong but that's probably not the problem. :)

My guess is the harvested directory doesn't exist. I *just* did the first [Deployment Dojo episode on harvesting](https://youtube.com/live/XcUmwegYClg) with more to follow over the next few weeks. It happens to cover this scenario (and more).

**Comment by MaroisConsulting** [comment]
Thanks for replying. I watched the video, but it's overwhelming.  I'm still having trouble. I'm pretty frustrated at this point.

1. Not sure what you mean by my extensions are wrong. I used the template files and all of their extensions are '.wxs'
2. You didn't answer the question: Isn't the harvesting supposed to produce an output file? I don't see one. With Wix4, how do I specify what that output file is? In v3, there is a '-out' param. How is this done in v4?
3. The directory I'm trying to harvest exists
4. No .msi is being created. 

I'm getting the compilation error
'WIX0091	Duplicate symbol 'WixComponentGroup:AppComponents''

I just threw together [this sample app](https://github.com/MaroisConsulting/Wix4SampleApp). It behaves the same. Can you please take a look and tell me what I'm doing wrong?

**Comment by robmen** [comment]
1. It was a joke. The first line in your question above is `Here's my Product.wsx:`. 😄  
2. I was pretty sure it only produces a file if something was harvested and your example showed that is correct. In the Deployment Dojo episode, I showed how the files are generated in the `obj` folder. The files end up automatically included.
3. I tried your sample. The harvested directory doesn't exist. The binlog shows heat skips harvesting if the directory doesn't exist.
4. The MSI was created for me.

Then I saw you updated the repo to fix the missing directory, and now everything works exactly as it should. You have two ComponentGroups with the same Id. The error messages are longer than just `'WIX0091 Duplicate symbol 'WixComponentGroup:AppComponents''` and they tell you exactly what the problem is.

The root problem in this case is that you had the wrong directory when harvesting.

**Reply by MaroisConsulting** [reply]
OK, I see now. 

I had a file called AppComponents in my installer project, and I was expecting THAT file to be filled with the target files. I just noticed that you said " I showed how the files are generated in the obj folder. The files end up automatically included."

Once I removed my copy of AppComponents it works as expected. 

Thank you!

**Reply by robmen** [reply]
You are lacking some fundamentals about how the WiX Toolset works. It will be more clear if you start at the beginning a bit. For example, I introduced harvesting in Episode 52 of the Deployment Dojo. You don't need _all_ the other 51 episodes before it but probably the first few will be enlightening.

Of course, you don't have to use the Deployment Dojo. I do that for fun. It's not designed to be a perfectly organized training. Tutorials and others may work for you.

**Reply by robmen** [reply]
I've never seen someone edit a comment inline as rapidly as you to change the meaning from the beginning to the end completely. It is very confusing responding to you.

**Reply by MaroisConsulting** [reply]
I apologize. I was trying different things and commenting as I was working.

I'll go back and watch Episode 52. I can't seem to find it. Can you post a link to it?

But I'm now getting an functional installer.

Thanks again

**Comment by BryanCrotaz** [comment]
I'm stuck with the same problem. I have gone through so many examples and docs and with a bit of help from @robmen I've worked out that the Registry is somehow tangled up with Harvesting - disabling harvesting the registry makes HarvestDirectory stop throwing errors. However the component group is still not found when the MSI build happens - however somehow I still get an MSI, but as it's only 32k, I assume it's empty!

@MaroisConsulting can you share the harvest directory settings you used? Here's mine... what did you do differently to make it work?

```xml
MyProject.wixproj
<Project Sdk="WixToolset.Sdk/4.0.2" TreatWarningsAsErrors="true">
  <PropertyGroup>
    <PreBuildEvent>
    </PreBuildEvent>
  </PropertyGroup>
	<ItemGroup>
		<HarvestDirectory Include="$(SolutionDir)AgentService\bin\Release\net7.0">
			<ComponentGroupName>AgentServiceFiles</ComponentGroupName>
			<DirectoryRefId>INSTALLFOLDER</DirectoryRefId>
			<SuppressRootDirectory>true</SuppressRootDirectory>
			<!-- dont know why but disabling harvesting registry keys makes HEAT5151 error go away here -->
			<SuppressRegistry>true</SuppressRegistry>
		</HarvestDirectory>

		<BindPath Include="$(SolutionDir)AgentService\bin\Release\net7.0" />
	</ItemGroup>
	<ItemGroup>
		<PackageReference Include="WixToolset.Heat" />
	</ItemGroup>
</Project>
```

```xml
package.wxs
<Wix xmlns="http://wixtoolset.org/schemas/v4/wxs" >
  <Package 
	  Name="AgentSetup"
	  Manufacturer="A Company"
	  Version="1.0.0.0"
	  UpgradeCode="68c96489-4459-4f0b-8c73-e4f6f8f8ddf1"
  >
    <MajorUpgrade DowngradeErrorMessage="!(loc.DowngradeError)" />

    <Feature Id="Main">
      <ComponentGroupRef Id="AgentServiceFiles" />
    </Feature>
  </Package>
</Wix>
```

In the diagnostic output I see this, but as HarvestDirectory was throwing loads of HEAT5151 errors until I disabled registry harvesting, I think these must be a red herring - it was clearly running and doing something.
```
Target "HarvestProjects" skipped, due to false condition; ( $(EnableProjectHarvesting) and ('@(HeatProject)' != '' or '@(HarvestProject)' != '') ) was evaluated as ( false and ('' != '' or '' != '') ).
Target "HarvestDirectory" skipped, due to false condition; ( '@(HarvestDirectory)' != '' ) was evaluated as ( '' != '' ).
Target "HarvestFile" skipped, due to false condition; ( '@(HarvestFile)' != '' ) was evaluated as ( '' != '' ).
```

I can't see any reference to setting a `HarvestDirectory` property in the docs so I don't know what this refers to.
